### PR TITLE
Add support for 'edge_cluster_id' specification in VDC

### PR DESCRIPTION
**Pull request by Didainius**
This PR adds an option to specify `edge_cluster_id` for NSX-T `vcd_org_vdc` resource. This is a prerequisite for DHCP management in NSX-T Isolated networks. Next step will be to align `vcd_nsxt_network_dhcp` resource with latest VCD features (issues #907 and #885 are also related to bumping up `vcd_nsxt_network_dhcp` as VCD functionality was very limited when we introduced it)
**Note.** Field `edge_cluster_id` always works for resources as only a provider user can create VDCs. The data source user might not always have access to read Edge Cluster configuration. For this reason the data source attempts to read value but ignores that error if it got an error in response.  (data source doc is also updated with information about behavior of this attribute)

Data source `vcd_nsxt_edge_cluster` used for Edge Cluster lookup is also improved to support better filtering - previously it was bound to VDC lookup only which wasn't suitable as there is no VDC yet when creating  `vcd_org_vdc`. Now it supports lookup by different filtering scopes as provided by VCD - `vdc_id`, `vdc_group_id` and `provider_vdc_id`. 

Also a new config parameter `nsxtEdgeCluster` is introduced to supply tests with existing NSX-T Edge Cluster name.

Screenshot of configuration location in UI VCD 10.4.0 
image

Partly related to:
* https://github.com/vmware/terraform-provider-vcd/issues/885

### CODE REVIEW

File: vcd/datasource_vcd_nsxt_edge_cluster.go [hunk 1]
```diff
@@ -18,9 +22,29 @@ func datasourceVcdNsxtEdgeCluster() *schema.Resource {
...
 			},
 			"vdc": {
-				Type:        schema.TypeString,
-				Optional:    true,
-				Description: "The name of VDC to use, optional if defined at provider level",
+				Type:          schema.TypeString,
+				Optional:      true,
+				Description:   "The name of VDC to use, optional if defined at provider level",
+				ConflictsWith: []string{"vdc_id", "vdc_group_id", "provider_vdc_id"},
+				Deprecated:    "This field is deprecated in favor of 'owner_id' which accepts IDs of VDC, VDC Group and Provider VDC",
```
Comment (adambarreiro) [inline vcd/datasource_vcd_nsxt_edge_cluster.go:29]:
```suggestion
				Deprecated:    "This field is deprecated in favor of 'vdc_id', 'vdc_group_id' and 'provider_vdc_id' which accepts IDs of VDC, VDC Group and Provider VDC respectively",
```

File: vcd/datasource_vcd_nsxt_edge_cluster.go [hunk 2]
```diff
@@ -18,9 +22,29 @@ func datasourceVcdNsxtEdgeCluster() *schema.Resource {
...
+			},
+			"vdc_id": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Description:   "ID of VDC, VDC Group or Provider VDC",
```
Comment (adambarreiro) [inline vcd/datasource_vcd_nsxt_edge_cluster.go:34]: Should it be
```suggestion
				Description:   "ID of VDC",
```
?

File: vcd/datasource_vcd_nsxt_edge_cluster.go [hunk 3]
```diff
@@ -18,9 +22,29 @@ func datasourceVcdNsxtEdgeCluster() *schema.Resource {
...
+				ConflictsWith: []string{"vdc", "vdc_group_id", "provider_vdc_id"},
+			},
+			"vdc_group_id": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Description:   "ID of VDC, VDC Group or Provider VDC",
```
Comment (adambarreiro) [inline vcd/datasource_vcd_nsxt_edge_cluster.go:40]: Should it be
```suggestion
				Description:   "ID of VDC Group",
```
?

File: vcd/datasource_vcd_nsxt_edge_cluster.go [hunk 4]
```diff
@@ -18,9 +22,29 @@ func datasourceVcdNsxtEdgeCluster() *schema.Resource {
...
+				ConflictsWith: []string{"vdc", "vdc_id", "provider_vdc_id"},
+			},
+			"provider_vdc_id": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Description:   "ID of VDC, VDC Group or Provider VDC",
```
Comment (adambarreiro) [inline vcd/datasource_vcd_nsxt_edge_cluster.go:46]: Should it be
```suggestion
				Description:   "ID of Provider VDC",
```
?

File: vcd/datasource_vcd_org_vdc.go [hunk 5]
```diff
@@ -240,5 +248,31 @@ func datasourceVcdOrgVdcRead(_ context.Context, d *schema.ResourceData, meta int
...
+
+	err = setDataSourceEdgeClusterData(d, adminVdc)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return nil
+}
+
+// setDataSourceEdgeClusterData is like setEdgeClusterData however it must handle
```
Comment (adambarreiro) [inline vcd/datasource_vcd_org_vdc.go:260]: It seems the function documentation is incomplete?

File: vcd/datasource_vcd_org_vdc.go [hunk 6]
```diff
@@ -240,5 +248,31 @@ func datasourceVcdOrgVdcRead(_ context.Context, d *schema.ResourceData, meta int
...
+func setDataSourceEdgeClusterData(d *schema.ResourceData, adminVdc *govcd.AdminVdc) error {
+	vdcNetworkProfile, err := adminVdc.GetVdcNetworkProfile()
+	if err != nil {
+		// Conciously ignoring this error and loggin it to output as it will most probably be
+		// insufficient rights that the user has. It will work with System user but might not work
+		// for users that got
```
Comment (adambarreiro) [inline vcd/datasource_vcd_org_vdc.go:266]: It seems also an incomplete comment

**Review by adambarreiro — COMMENTED** [review]
First review, found some little things to clarify:

File: vcd/datasource_vcd_nsxt_edge_cluster.go (hunk 4 above)
Comment (Didainius) [reply vcd/datasource_vcd_nsxt_edge_cluster.go:46]: Yes! I messed this up from initial mockups of code.

File: vcd/datasource_vcd_nsxt_edge_cluster.go (hunk 3 above)
Comment (Didainius) [reply vcd/datasource_vcd_nsxt_edge_cluster.go:40]: Yes! I messed this up from initial mockups of code.

File: vcd/datasource_vcd_nsxt_edge_cluster.go (hunk 2 above)
Comment (Didainius) [reply vcd/datasource_vcd_nsxt_edge_cluster.go:34]: Yes! I messed this up from initial mockups of code.

File: .changes/v3.8.0/921-improvements.md [hunk 7]
```diff
@@ -0,0 +1,4 @@
+* Data source `vcd_nsxt_edge_cluster` deprecated `vdc` field in favor of 3 new fields to define
```
Comment (lvirbalas) [inline .changes/v3.8.0/921-improvements.md:1]: This is an ask for the PR description:

>  This is a prerequisite for DHCP management in NSX-T Isolated networks

Do we have an issue we could link in the PR description?

File: website/docs/d/org_vdc.html.markdown [hunk 8]
```diff
@@ -38,5 +38,10 @@ The following arguments are supported:
 
 ## Attribute reference
 
-All attributes defined in [organization VDC resource](/providers/vmware/vcd/latest/docs/resources/org_vdc#attribute-reference) are supported.
+* `edge_cluster_id` - (*v3.8+*, *VCD 10.3+*) An ID of NSX-T Edge Cluster which should provide vApp
+  Networking Services or DHCP for isolated networks. This value **might be unavailable** in data
```
Comment (lvirbalas) [inline website/docs/d/org_vdc.html.markdown:42]: Maybe emphasize here?

```suggestion
  Networking Services or DHCP for Isolated Networks. This value **might be unavailable** in data
```

**Review by lvirbalas — APPROVED** [review]
LGTM, and nice move in favor of vdc_id, vdc_group_id and provider_vdc_id fields!

Just one ask for PR description.

File: .changes/v3.8.0/921-improvements.md (hunk 7 above)
Comment (Didainius) [reply .changes/v3.8.0/921-improvements.md:1]: > Do we have an issue we could link in the PR description?

It was not exactly the same, but very much related to catching up with NSX-T DHCP (vcd_nsxt_network_dhcp) functionality for Org networks ->  https://github.com/vmware/terraform-provider-vcd/issues/907 and https://github.com/vmware/terraform-provider-vcd/issues/885

File: website/docs/d/org_vdc.html.markdown (hunk 8 above)
Comment (Didainius) [reply website/docs/d/org_vdc.html.markdown]: Will not do harm - Improved.

**Review by adambarreiro — APPROVED** [review]

**Pull request merged**